src/main/java/io/cdap/plugin/gcp/bigtable/source/BigtableSource.java

```diff
@@ -0,0 +1,249 @@
+        recordBuilder.set(fieldName, value);
+      } else {
+        HBaseColumn hBaseColumn = HBaseColumn.fromFullName(fieldName);
+        byte[] valueBytes =
+          result.getValue(Bytes.toBytes(hBaseColumn.getFamily()), Bytes.toBytes(hBaseColumn.getQualifier()));
+        Object value = convertBytesToField(valueBytes, fieldSchema);
+        recordBuilder.set(fieldName, value);
+      }
+    }
+    return recordBuilder.build();
+  }
+
+  private Object convertBytesToField(byte[] bytes, Schema fieldSchema) {
+    if (fieldSchema.isNullable()) {
+      if (null == bytes) {
+        return null;
+      }
+      return convertBytesToField(bytes, fieldSchema.getNonNullable());
+    }
+    switch (fieldSchema.getType()) {
+      case NULL:
+        return null;
+      case STRING:
+        return Bytes.toString(bytes);
+      case BYTES:
+        return bytes;
+      case INT:
+        return Bytes.toInt(bytes);
+      case LONG:
+        return Bytes.toLong(bytes);
+      case FLOAT:
+        return Bytes.toFloat(bytes);
+      case DOUBLE:
+        return Bytes.toDouble(bytes);
+      case BOOLEAN:
+        return Bytes.toBoolean(bytes);
+      default:
+        throw new UnexpectedFormatException("Field type '" + fieldSchema.getType() + "' is not supported");
+    }
+  }
+
+  private void validateOutputSchema(Configuration configuration) {
+    TableName tableName = TableName.valueOf(config.table);
+    try (Connection connection = BigtableConfiguration.connect(configuration);
+         Table table = connection.getTable(tableName)) {
+      Set<String> existingFamilies = table.getTableDescriptor()
+        .getFamiliesKeys()
+        .stream()
+        .map(Bytes::toString)
+        .collect(Collectors.toSet());
+      for (HBaseColumn hBaseColumn : config.getRequestedColumns()) {
+        if (!existingFamilies.contains(hBaseColumn.getFamily())) {
+          throw new InvalidConfigPropertyException(
+            String.format("Column family '%s' does not exist", hBaseColumn.getFamily()),
+            BigtableSourceConfig.SCHEMA
+          );
+        }
+      }
+    } catch (IOException e) {
+      throw new InvalidStageException("Failed to connect to Bigtable", e);
+    }
+  }
+
+  private void emitLineage(BatchSourceContext context) {
+    LineageRecorder lineageRecorder = new LineageRecorder(context, config.referenceName);
+    lineageRecorder.createExternalDataset(config.getSchema());
+
+    List<Schema.Field> fields = Objects.requireNonNull(config.getSchema()).getFields();
+    if (null != fields) {
+      List<String> fieldNames = fields.stream()
+        .map(Schema.Field::getName)
+        .collect(Collectors.toList());
+      lineageRecorder.recordRead("Read", "Read from Bigtable.", fieldNames);
+    }
+  }
+
+  private Scan getConfiguredScanForJob() throws IOException {
+    Scan s = new Scan();
+    if (null != config.scanTimeRangeStart && null != config.scanTimeRangeStop) {
```

Comment:
does setTimeRange require non-null values for both the time range need both start and end? Logically it seems like only one should be required. If start is not given, there is no start time (0). If end is not given, there is no end time (long max value)

Reply:
fixed